pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/TableRebalancerAdminToolClusterIntegrationTest.java

```diff
@@ -0,0 +1,877 @@
+    final int numSegments = 4;
+    final int numReplicas = 3;
+    final int numAdditionalServers = 3;
+    final String tableName = getTableName() + "LLNOCONSUMING";
+
+    try {
+      startFakeServers(NUM_INITIAL_SERVERS, CommonConstants.Helix.DEFAULT_SERVER_NETTY_PORT, true);
+      startKafka();
+      setUpRealtimeTable(null, numReplicas, true, tableName);
+
+      // init stats
+      _segmentStateTransitionStats = new SegmentStateTransitionStats();
+
+      // write ideal state
+      createIdealStateForRealtime(numSegments, numReplicas, tableName, true);
+
+      // add additional servers to trigger the rebalance strategy,
+      // otherwise rebalance will be a NO-OP
+      startFakeServers(numAdditionalServers, CommonConstants.Helix.DEFAULT_SERVER_NETTY_PORT + NUM_INITIAL_SERVERS, true);
+
+      // rebalance
+      final PinotTableRebalancer tableRebalancer = new PinotTableRebalancer(ZKSTR, getHelixClusterName(), false, true, false, 2);
+      tableRebalancer.rebalance(tableName, "REALTIME");
+      final TableRebalancer.RebalancerStats stats = tableRebalancer.getRebalancerStats();
+
+      // verify the algorithm stats on how we moved from
+      // current to target ideal state
+      Assert.assertEquals(0, stats.getDryRun());
+      Assert.assertEquals(stats.getUpdatestoIdealStateInZK(), 2);
+      Assert.assertEquals(stats.getDirectUpdatesToSegmentInstanceMap(), 7);
+      Assert.assertEquals(stats.getIncrementalUpdatesToSegmentInstanceMap(), 1);
+      Assert.assertEquals(stats.getNumSegmentMoves(), 3);
+
+      // as part of rebalancing, host2 lost segment1 and segment3,
+      // host1 and host3 lost segment2 and segment4 -- so 6
+      // transitions from ON to OFF and OFF to DROP
+      Assert.assertEquals(_segmentStateTransitionStats.offFromOn, 3);
+      Assert.assertEquals(_segmentStateTransitionStats.dropFromOff, 3);
+    } finally {
+      stopFakeServers();
+      stopKafka();
+    }
+  }
+
+  /**
+   * Test for realtime table in no downtime mode
+   * with HLC consumer
+   *
+   * Scenario: 4 segments 3 replicas
+   *
+   * Current ideal state
+   *
+   * segment__0__0: {host1:online, host2:online, host3:online}
+   * segment__0__1: {host1:consuming, host2:consuming, host3:consuming}
+   * segment__1__0: {host1:online, host2:online, host3:online}
+   * segment__1__1: {host1:consuming, host2:consuming, host3:consuming}
+   *
+   * We add 3 additional hosts: host4, host5 and host6 and check that
+   * rebalancer does not do anything for realtime table with
+   * HLC consumer
+   *
+   * @throws Exception
+   */
+  @Test
+  public void testHLCRealtimeTable() throws Exception {
+    final int numSegments = 4;
+    final int numReplicas = 3;
+    final int numAdditionalServers = 3;
+    final String tableName = getTableName() +"HLC";
+
+    try {
+      startFakeServers(NUM_INITIAL_SERVERS, CommonConstants.Helix.DEFAULT_SERVER_NETTY_PORT, true);
+      startKafka();
```

Comment:
things may run faster if you startKafka and just keep it around and stopKafka() at the end. We don't consume any messages, so it is ok to just keep it running throughout.

Reply:
Done